contracts/Zap/ZapMint.sol

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.13;
+
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@uniswap/v2-periphery/contracts/interfaces/IUniswapV2Router02.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "../BdStable/Pools/BdStablePool.sol";
+import "../BSM/Investors/AMM/AMMInvestorsHelper.sol";
+import "hardhat/console.sol";
+
+//todo
+//use PausableUpgradeable?
+//optimise gas usage
+//use AMMInvestorsHelper (test for sovryn router)
+contract ZapMint is OwnableUpgradeable, ReentrancyGuard {
+    using SafeERC20 for IERC20;
+
+    bool public zapMintPaused;
+
+    mapping(address => bool) private mapZapMintSupportedTokens;
+    address[] public zapMintSupportedTokens;
+
+    function initialize() external initializer {
+        __Ownable_init();
+        zapMintPaused = false;
+    }
+
+    function isSupportedForMinting(address _address) public view returns (bool) {
+        return mapZapMintSupportedTokens[_address];
+    }
+
+    function addTokenSupportForMinting(address token) public onlyOwner {
+        if (!mapZapMintSupportedTokens[token]) {
+            mapZapMintSupportedTokens[token] = true;
+            zapMintSupportedTokens.push(token);
+
+            emit AddTokenSupportForMinting(token);
+        }
+    }
+
+    function removeTokenSupportFromMinting(uint256 index) external onlyOwner {
+        address token = zapMintSupportedTokens[index];
+        if (mapZapMintSupportedTokens[token]) {
+            mapZapMintSupportedTokens[token] = false;
+            zapMintSupportedTokens[index] = zapMintSupportedTokens[zapMintSupportedTokens.length - 1];
+            zapMintSupportedTokens.pop();
+
+            emit RemoveTokenSupportFromMinting(token);
+        }
+    }
+
+    function toggleZapMinting() external onlyOwner {
+        zapMintPaused = !zapMintPaused;
+    }
+
+    function mintUniswapRouterOnly(
+        uint256 bdxInMax,
```

Comment:
We do for now but I think we should not, I'll remove that and make both collateral and bdx come from result of swapping input token. As far as I remember the Smart View has to operate only on input token.
We might in smart view:
* send bdx or collateral from user if there is enough of them in the wallet
* invoke normal mint in FE if wallet has enought of both collateral and BDX

But I think it's not very transparent for the user (imagine choosing to mint from XUSD, filling up the field only to notice your RBTC and/or BDX has gone instead of XUSD).

Reply:
![image](https://user-images.githubusercontent.com/91085805/172216602-622b545b-3161-483f-a7ee-859bb450571f.png)

We don't notify the user either that we are going to use some of his collateral/bdx in current wireframes.

Reply:
Changes implementation to NOT provide bdx to zap mint - all collateral and BDX choung come as a result of swapping input token.